File: src/Compilers/VisualBasic/Portable/TypeExtensions.vb
' -----------------------------------------------------------------------
' <copyright file="TypeExtensions.vb" company="Altemiq">
' Copyright (c) Altemiq. All rights reserved.
' </copyright>
' -----------------------------------------------------------------------

Public Module TypeExtensions
    ''' <summary>
    ''' Converts the type to the <see cref="Syntax.TypeSyntax"/>.
    ''' </summary>
    ''' <param name="parameters">The type parameters.</param>
    ''' <returns>The type syntax.</returns>
    ''' <exception cref="ArgumentOutOfRangeException">The parameters are the wrong length.</exception>
    <Runtime.CompilerServices.Extension>
    Public Function ToTypeSyntax(type As Type, ParamArray parameters() As Syntax.TypeSyntax) As Syntax.NameSyntax
        If Not type.IsGenericType Then
            If type.FullName IsNot Nothing Then
                Return SyntaxFactory.ParseName(type.FullName)
            End If

            Throw New InvalidOperationException
        End If

        Dim index = type.Name.IndexOf("`"c)
        Dim count = Integer.Parse(type.Name.Substring(index + 1), Globalization.CultureInfo.InvariantCulture)

        Dim parameterList = SyntaxFactory.SeparatedList(parameters)
        If count <> parameterList.Count Then
            Throw New ArgumentOutOfRangeException(NameOf(parameters))
        End If

        Dim genericName = SyntaxFactory.GenericName(SyntaxFactory.Identifier(type.Name.Substring(0, index)), SyntaxFactory.TypeArgumentList(parameterList))
        Return If(
            type.Namespace IsNot Nothing,
            SyntaxFactory.QualifiedName(SyntaxFactory.ParseName(type.Namespace), genericName),
            DirectCast(genericName, Syntax.NameSyntax))
    End Function
End Module

File: src/Compilers/CSharp/Portable/TypeExtensions.cs
// -----------------------------------------------------------------------
// <copyright file="TypeExtensions.cs" company="Altemiq">
// Copyright (c) Altemiq. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace Microsoft.CodeAnalysis.CSharp;

#pragma warning disable RCS1263, SA1101

/// <summary>
/// The type extensions.
/// </summary>
public static class TypeExtensions
{
    /// <content>
    /// The <see cref="Type"/> extensions.
    /// </content>
    /// <param name="type">The type.</param>
    extension(Type type)
    {
        /// <summary>
        /// Converts the type to the <see cref="Syntax.TypeSyntax"/>.
        /// </summary>
        /// <param name="parameters">The type parameters.</param>
        /// <returns>The type syntax.</returns>
        /// <exception cref="ArgumentOutOfRangeException">The parameters are the wrong length.</exception>
        public Syntax.NameSyntax ToTypeSyntax(IEnumerable<Syntax.TypeSyntax> parameters)
        {
            if (!type.IsGenericTypeDefinition)
            {
                return type.FullName is { } fullName
                    ? SyntaxFactory.QualifiedName(fullName)
                    : throw new InvalidOperationException();
            }

            var index = type.Name.IndexOf('`');
            var count = int.Parse(type.Name.Substring(index + 1), System.Globalization.CultureInfo.InvariantCulture);

            var parameterList = SyntaxFactory.SeparatedList(parameters);
            if (count != parameterList.Count)
            {
                throw new ArgumentOutOfRangeException(nameof(parameters));
            }

            var name = type.Name.Substring(0, index);
            var genericName = SyntaxFactory.GenericName(SyntaxFactory.Identifier(name), SyntaxFactory.TypeArgumentList(parameterList));
            return type is { Namespace: { } n }
                ? SyntaxFactory.QualifiedName(SyntaxFactory.QualifiedName(n), genericName)
                : genericName;
        }
    }
}